lib/report/collector.go

```diff
@@ -0,0 +1,179 @@
+type FileWriter interface {
+	NewWriter(name string) (io.WriteCloser, error)
+}
+
+// NewWriter creates a new writer to write to a file with the specified name
+func (r FileWriterFunc) NewWriter(name string) (io.WriteCloser, error) {
+	return r(name)
+}
+
+// FileWriterFunc is a functional wrapper for NamedWriter
+type FileWriterFunc func(name string) (io.WriteCloser, error)
+
+// Collector defines an interface to collect diagnostic information
+type Collector interface {
+	// Collect collects diagnostics using CommandRunner and serializes
+	// them using specified Writer
+	Collect(context.Context, FileWriter, utils.CommandRunner) error
+}
+
+// Collectors is a list of Collectors
+type Collectors []Collector
+
+// Collect implements Collector for a list of Collectors
+func (r Collectors) Collect(ctx context.Context, reportWriter FileWriter, runner utils.CommandRunner) error {
+	var errors []error
+	for _, collector := range r {
+		err := collector.Collect(ctx, reportWriter, runner)
+		if err != nil {
+			errors = append(errors, err)
+		}
+	}
+	return trace.NewAggregate(errors...)
+}
+
+// Cmd creates a new Command with the given name and command line
+func Cmd(name string, args ...string) Command {
+	cmd := args[0]
+	args = args[1:]
+	return Command{name: name, cmd: cmd, args: args}
+}
+
+// Self returns a reference to this binary.
+// name names the resulting output file.
+// args is the list of optional command line arguments
+func Self(name string, args ...string) Command {
+	return Command{name: name, cmd: utils.Exe.Path, args: args}
+}
+
+// Command defines a generic command with a name and a list of arguments
+type Command struct {
+	name string
+	cmd  string
+	args []string
+}
+
+// Collect implements Collector for this Command
+func (r Command) Collect(ctx context.Context, reportWriter FileWriter, runner utils.CommandRunner) error {
+	w, err := reportWriter.NewWriter(r.name)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer w.Close()
+
+	args := []string{r.cmd}
+	args = append(args, r.args...)
+	return runner.RunStream(ctx, w, args...)
+}
+
+// Script creates a new script collector
+func Script(name, script string) ScriptCollector {
+	return ScriptCollector{name: name, script: script}
+}
+
+// Collect implements Collector using a bash script
+func (r ScriptCollector) Collect(ctx context.Context, reportWriter FileWriter, runner utils.CommandRunner) error {
+	args := []string{"/bin/bash", "-c", r.script}
+	w, err := reportWriter.NewWriter(r.name)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer w.Close()
+
+	return runner.RunStream(ctx, w, args...)
+}
+
+// ScriptCollector is a convenience Collector to execute bash scripts
+type ScriptCollector struct {
+	name   string
+	script string
+}
+
+func tarball(pattern string) string {
+	return fmt.Sprintf(`
+#!/bin/bash
+/bin/tar cz --ignore-failed-read --ignore-command-error -f /dev/stdout -C / $(readlink -e %v) -P 2> /dev/null
```

Comment:
Hm, is this script run over ssh somewhere so it's not possible to use `archive`package instead of running bash? And where does the `pattern` come from? Just thinking about the possibility of injection attack.

Reply:
`pattern` is hardcoded and only used internally.
I will create a separate issue to refactor this using archive package.